feat: provide debugging capabilities

GregoryAlbouy opened the pull request:
<!-- IMPORTANT: Don't forget to link the issue(s) once the PR created! -->

## Description

Not much to add, does pretty much exactly what the issue suggests.

<!--
    Describe briefly what this PR does, if the issue description is not enough.
    Add any information that could be relevant for the reviewers.
-->

## Changes

- Implemented a `struct server` to avoid global variables for testability purposes

<!-- Optional: mention here indirect changes impacted by the PR -->

## Notes

<!-- Optional: additional notes than can help understanding the implementation -->

GregoryAlbouy committed:
Provide debugging capabilities via CLI

GregoryAlbouy committed:
Provide debugging capabilities via endpoint /debug

GregoryAlbouy committed:
Avoid global variables for testability

- Implement struct server with mutex and totalRequest attached
- Update unit tests

GregoryAlbouy committed:
Write unit tests for server.handleDebug

moreirathomas commented on main.go:
```diff
@@ -15,8 +15,13 @@ func main() {
-	fmt.Printf("http://localhost:%s\n", *port)
-	log.Fatal(http.ListenAndServe(addr, handler))
+	fmt.Printf("http://localhost%s\n", addr)
+	go func() {
+		log.Fatal(http.ListenAndServe(addr, handler))
+	}()
+
+	log.Fatal(server.listenInput())
```
I'd actually invert these, `listenInput()` would go in a goroutine because it's not the _main_ process to run.
`http.ListenAndServe()` would still do its forever loop as it always does.

moreirathomas commented on handle.go:
```diff
@@ -8,14 +8,29 @@ import (
 
+type urlpath string
+
+const urlpathDebug urlpath = "/debug"
```
Move it to `const` block below (line 18) ?

GregoryAlbouy replied:
To me it makes sense to group these constants by type because they serve as an enum here.

moreirathomas replied:
Oh i did not think about that, all right then :)

moreirathomas commented:
If we're adding features to the server, I would be to start actually buidling a package thats is not `main`.

Something to use like:

```go
// main.go
package main

import "github.com/benchttp/cobaye/server"

func main() {
	// ...

	s := server.New() // struct server.Server only exposing 2 methods
	
	go func() {
		log.Fatal(s.ListenStdinInput())
	}()

	addr := ":" + *port
	log.Fatal(http.ListenAndServe(addr, s.Handle))
}
```

GregoryAlbouy committed:
Move server into its own package

GregoryAlbouy committed:
Refactor server.Server

- make it a http.Handler
- use atomic instead of mutexes for concurrent mutations
- log non-critical errors in Server.listenStdin
- call ListenAndServe in main goroutine
- update unit tests accordingly

GregoryAlbouy committed:
Update docs

c1-ra commented on internal/server/handle_test.go:
```diff
@@ -0,0 +1,109 @@
+
+	t.Run("request with delay param", func(t *testing.T) {
+		const delay = 100
+		const expmin, expmax = delay * time.Millisecond, (delay + 10) * time.Millisecond
```
This one fails on my cpu _sometimes_ (when it happens, it is only by 0.01s: at 0.11s).

~~EDIT: eventually, I am not sure the duration is relevant for each request.
What we want to check if especially that the requests are sent and that we get a response, right?
If it feels necessary, we could add a timeout to the each test maybe?~~

Is `check.Duration.InRange` checking that the response has been received between `expmin` and `expmax`, since the method has been launched (and not since the request has been sent)?

GregoryAlbouy replied:
In short: only the handler's execution time

Technically, no request is sent: handlers are regular functions and we can unit test them by calling them with specific arguments (one of them being a request object) and see the results.
`drykit-go/testx` is an amazingly well thought lib, as a consequence it only measures the execution time of the handler (just like you would for any regular function):
https://github.com/drykit-go/testx/blob/2f52657f1d91930702d638506c3c307f2f9acc36/runner_httphandler.go#L84-L86

c1-ra replied:
Ok thanks!
What an amazingly well thought lib, indeed 😄 
The PR is ok for me as it is then.

c1-ra commented on internal/server/handle_test.go:
```diff
@@ -0,0 +1,109 @@
+
+	t.Run("request with fib param", func(t *testing.T) {
+		const fib = 35
+		const expmin, expmax = 30 * time.Millisecond, 80 * time.Millisecond
```
Do we need to check the duration for this one?
I feel like checking that we receive a 200 status code is enough.

c1-ra reviewed (commented):
Nice, seems ok to me, except for the question about the duration checks.

GregoryAlbouy committed:
Rename file handle_test.go -> server_test.go

GregoryAlbouy commented:
Rebased with conflicts, no other changes.
